ReadMe
libmots
-------

libmots library allows you to display additional debug information on
the screen. It is very useful when you cannot keep active adb
connection, e.g. while debugging application that changes USB states. libmots
library has an advantage over toasts - history is available. This can be very
useful for longer scenarios. You can start logging at any time you want and
clear already collected messages.

You can find screenshots presenting use of 'Hello libmots' application
inside this directory: tests/monkeyrunner-tests/reference-screenshots

USAGE
-----

To start, you need to add libmots library to your project and add this
line to your main project.properties file:

	manifestmerger.enabled=true

This is simplest part you need to add to your source code:

	...
	import net.wiktorlawski.messageonthescreen.MessageOnTheScreen;
	...
	MessageOnTheScreen mots =
		MessageOnTheScreen.getInstance(activity, false);
	mots.setText(activity, "Hello libmots!");

SAMPLES & TESTS
---------------

Library comes with both sample application (samples/ directory) and tests
(tests/ directory). monkeyrunner tests use emulator which you can prepare
running this command:

	android create avd -n android-1.6 -t android-4

(without creation of custom hardware profile).

To run monkeyrunner tests, just start run-tests.bat from
tests/monkeyrunner-tests directory.

LICENSE
-------

The MIT License (MIT)


ChangeLog
ver 1.0
	Made 'shared element' movable and removable from UI.
	Added option to clear debug window.

ver 0.3
	Added interface to add new debug messages to debug window.

ver 0.2
	Added optional auto-hiding for 'shared element'.

ver 0.1
	Added 'shared element'.
	Added debug window available through 'shared element'.


tests/monkeyrunner-tests/helpers.py
#
# The MIT License (MIT)
#
# Copyright (c) 2014 Wiktor Lawski <wiktor.lawski@gmail.com>
#
# Permission is hereby granted, free of charge, to any person obtaining a copy
# of this software and associated documentation files (the "Software"), to deal
# in the Software without restriction, including without limitation the rights
# to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
# copies of the Software, and to permit persons to whom the Software is
# furnished to do so, subject to the following conditions:
#
# The above copyright notice and this permission notice shall be included in
# all copies or substantial portions of the Software.
#
# THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
# IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
# FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
# AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
# LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
# OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
# THE SOFTWARE.
#

#
# This file provides a set of functions that are common for most of the
# monkeyrunner tests.
#

import os

from com.android.monkeyrunner import MonkeyDevice, MonkeyRunner
from java.lang import NullPointerException

actualDir = "actual-screenshots"
defaultEmulator = "emulator-5554"
landscapeRect = (0, 0, 440, 800)
numberOfTabs = 7
portraitRect = (0, 35, 480, 765)
referenceDir = "reference-screenshots"
tabSize = 8

package = "net.wiktorlawski.hellolibmots"
activity = package + ".HelloLibmotsActivity"
runComponent = package + "/" + activity

shortWaitTime = 0.5
waitTime = 2.5
longWaitTime = 7.0

binLscapeX = 400
binLscapeY = 430
binX = 240
binY = 750
sharedElemX = 405
sharedElemY = 110

# Compares actual screenshot with the reference one and based on that prints
# result of test from the script that called this function.
def checkResult(result, currentDir, calledScript):
    fullScriptName = os.path.basename(calledScript)
    scriptName = os.path.splitext(fullScriptName)[0]
    referenceFile = currentDir + "/" + referenceDir + "/" + scriptName + ".png"
    actualFile = currentDir + "/" + actualDir + "/" + scriptName + ".png"

    # 1 character is added to script name (':')
    indentSize = numberOfTabs - ((len(scriptName) + 1) / tabSize)

    prefix = ""
    for i in range(0, indentSize):
        prefix = prefix + "\t"

    try:
        if (result.sameAs(MonkeyRunner.loadImageFromFile(referenceFile), 0.999)):
            print scriptName + ":" + prefix + "Passed"
        else:
            result.writeToFile(actualFile)
            print scriptName + ":" + prefix + "Failed"
    except(NullPointerException):
        result.writeToFile(actualFile)
        print "Reference file (" + referenceFile + ") missing!"
        print scriptName + ":" + prefix + "Failed"

# Starts 'Hello libmots' application and creates shared element using
# "getInstance" button
def createSharedElement(device, showing):
    device.startActivity(component=runComponent)
    MonkeyRunner.sleep(waitTime)

    # Tick checkbox, if necessary
    if (showing):
        device.touch(30, 110, MonkeyDevice.DOWN_AND_UP)
        MonkeyRunner.sleep(waitTime)

    # Touch 'getInstance' button
    device.touch(270, 115, MonkeyDevice.DOWN_AND_UP)
    MonkeyRunner.sleep(waitTime)

# Returns device object
def getDevice():
    return MonkeyRunner.waitForConnection(10, defaultEmulator)

# Presses specified point for enough period of time to trigger long press
# behavior
def longPress(device, x, y):
    device.touch(x, y, MonkeyDevice.DOWN)
    waitForLongPress()

# Moves "finger" away from the screen
def moveFingerAway(device, x, y):
    device.touch(x, y, MonkeyDevice.UP)
    MonkeyRunner.sleep(waitTime)

# Moves shared element (moving "finger" away from the screen is optional)
def moveSharedElement(device, x, y, deltaX, deltaY, toMoveFingerAway=True):
    if (toMoveFingerAway):
        device.touch(x, y, MonkeyDevice.DOWN)

    device.touch(x + deltaX, y + deltaY, MonkeyDevice.MOVE)

    if (toMoveFingerAway):
        moveFingerAway(device, x + deltaX, y + deltaY)

    MonkeyRunner.sleep(shortWaitTime)

# Presses Back button
def pressBack(device):
    device.press("KEYCODE_BACK", MonkeyDevice.DOWN_AND_UP)
    MonkeyRunner.sleep(waitTime)

# Presses Home button
def pressHome(device):
    device.press("KEYCODE_HOME", MonkeyDevice.DOWN_AND_UP)
    MonkeyRunner.sleep(waitTime)

# Starts activity that sets device orientation to landscape
def setLandscape(device):
    package = "net.wiktorlawski.setorientation"
    activity = package + ".SetLandscape"
    runComponent = package + "/" + activity
    device.startActivity(component=runComponent)
    MonkeyRunner.sleep(longWaitTime)

# Starts activity that sets device orientation to portrait
def setPortrait(device):
    package = "net.wiktorlawski.setorientation"
    activity = package + ".SetPortrait"
    runComponent = package + "/" + activity
    device.startActivity(component=runComponent)
    MonkeyRunner.sleep(longWaitTime)

# Starts 'Hello libmots' application
def startHelloLibmots(device):
    device.startActivity(component=runComponent)
    MonkeyRunner.sleep(waitTime)

# Toggles "startShowing" checkbox
def toggleStartShowing(device):
    device.touch(30, 110, MonkeyDevice.DOWN_AND_UP)

# Touches "addMessage" button
def touchAddMessage(device):
    device.touch(185, 255, MonkeyDevice.DOWN_AND_UP)
    MonkeyRunner.sleep(shortWaitTime)

# Touches "Cancel" button in debug window
def touchCancel(device):
    device.touch(350, 485, MonkeyDevice.DOWN_AND_UP)
    MonkeyRunner.sleep(waitTime)

# Touches "Clear" button in debug window
def touchClear(device):
    device.touch(130, 485, MonkeyDevice.DOWN_AND_UP)
    MonkeyRunner.sleep(waitTime)

# Touches "getInstance" button
def touchGetInstance(device):
    device.touch(270, 115, MonkeyDevice.DOWN_AND_UP)
    MonkeyRunner.sleep(waitTime)

# Touches new message edit text
def touchNewMessageEditText(device):
    device.touch(25, 215, MonkeyDevice.DOWN_AND_UP)
    MonkeyRunner.sleep(shortWaitTime)

# Touches "setText" button
def touchSetText(device):
    device.touch(50, 285, MonkeyDevice.DOWN_AND_UP)
    MonkeyRunner.sleep(shortWaitTime)

# Touches shared element (using default shared element position)
def touchSharedElement(device):
    device.touch(sharedElemX, sharedElemY, MonkeyDevice.DOWN_AND_UP)
    MonkeyRunner.sleep(longWaitTime)

# Waits for enough period of time to trigger long press behavior
def waitForLongPress():
    MonkeyRunner.sleep(waitTime)


tests/monkeyrunner-tests/t0000-no-operation.py
#
# The MIT License (MIT)
#
# Copyright (c) 2014 Wiktor Lawski <wiktor.lawski@gmail.com>
#
# Permission is hereby granted, free of charge, to any person obtaining a copy
# of this software and associated documentation files (the "Software"), to deal
# in the Software without restriction, including without limitation the rights
# to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
# copies of the Software, and to permit persons to whom the Software is
# furnished to do so, subject to the following conditions:
#
# The above copyright notice and this permission notice shall be included in
# all copies or substantial portions of the Software.
#
# THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
# IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
# FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
# AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
# LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
# OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
# THE SOFTWARE.
#

#
# This test checks that there is no visual change just by installing
# application that uses libmots library.
#

import inspect
import os
import sys

from com.android.monkeyrunner import MonkeyRunner, MonkeyDevice

currentDir = os.environ["LIBMOTS_TEST_DIR"]
sys.path.insert(0, currentDir)

import helpers

calledScript = inspect.getfile(inspect.currentframe())
device = helpers.getDevice()

# Unlock the screen, if necessary (run only for this test)
device.press("KEYCODE_MENU", MonkeyDevice.DOWN_AND_UP)
MonkeyRunner.sleep(helpers.waitTime)

result = device.takeSnapshot().getSubImage(helpers.portraitRect)
helpers.checkResult(result, currentDir, calledScript)


tests/monkeyrunner-tests/t0001-start-sample.py
#
# The MIT License (MIT)
#
# Copyright (c) 2014 Wiktor Lawski <wiktor.lawski@gmail.com>
#
# Permission is hereby granted, free of charge, to any person obtaining a copy
# of this software and associated documentation files (the "Software"), to deal
# in the Software without restriction, including without limitation the rights
# to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
# copies of the Software, and to permit persons to whom the Software is
# furnished to do so, subject to the following conditions:
#
# The above copyright notice and this permission notice shall be included in
# all copies or substantial portions of the Software.
#
# THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
# IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
# FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
# AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
# LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
# OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
# THE SOFTWARE.
#

#
# This test checks that starting application that has not used libmots library
# yet, does not create any additional visual content.
#

import inspect
import os
import sys

from com.android.monkeyrunner import MonkeyRunner, MonkeyDevice

currentDir = os.environ["LIBMOTS_TEST_DIR"]
sys.path.insert(0, currentDir)

import helpers

calledScript = inspect.getfile(inspect.currentframe())
device = helpers.getDevice()

helpers.startHelloLibmots(device)

result = device.takeSnapshot().getSubImage(helpers.portraitRect)
helpers.checkResult(result, currentDir, calledScript)


src/net/wiktorlawski/messageonthescreen/DebugWindow.java
/*
 * The MIT License (MIT)
 *
 * Copyright (c) 2014 Wiktor Lawski <wiktor.lawski@gmail.com>
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in
 * all copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
 * THE SOFTWARE.
 */

package net.wiktorlawski.messageonthescreen;

import java.util.ArrayList;

import android.app.AlertDialog;
import android.content.Context;
import android.content.DialogInterface;
import android.content.DialogInterface.OnClickListener;
import android.view.WindowManager;
import android.widget.ArrayAdapter;

/**
 * This class prepares dialog that presents gathered debug messages.
 */
/* package */ class DebugWindow {
	SharedElementService service;

	/* package */ DebugWindow(Context context,
			ArrayList<String> debugMessages, SharedElementService service) {
		this.service = service;

		AlertDialog.Builder builder = new AlertDialog.Builder(context);
		ArrayAdapter<String> adapter =
				new ArrayAdapter<String>(context, R.layout.debug_message);

		for (String s : debugMessages) {
			adapter.add(s);
		}

		builder.setAdapter(adapter, new DebugMessageClickListener());
		builder.setNegativeButton(R.string.cancel,
				new DebugWindowCancelListener());
		builder.setNeutralButton(R.string.clear,
				new DebugWindowClearListener());

		AlertDialog dialog = builder.create();
		dialog.getWindow()
			.setType(WindowManager.LayoutParams.TYPE_SYSTEM_ALERT);

		dialog.show();
	}

	private class DebugWindowCancelListener implements OnClickListener {

		@Override
		public void onClick(DialogInterface dialog, int which) {
			dialog.dismiss();
		}
	}	

	private class DebugWindowClearListener implements OnClickListener {

		@Override
		public void onClick(DialogInterface dialog, int which) {
			service.clearDebugMessages();
		}
	}

	private class DebugMessageClickListener implements OnClickListener {

		@Override
		public void onClick(DialogInterface dialog, int which) {
			/* Nothing */
		}		
	}
}


src/net/wiktorlawski/messageonthescreen/MessageOnTheScreen.java
/*
 * The MIT License (MIT)
 *
 * Copyright (c) 2014 Wiktor Lawski <wiktor.lawski@gmail.com>
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in
 * all copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
 * THE SOFTWARE.
 */

package net.wiktorlawski.messageonthescreen;

import android.app.Activity;
import android.content.Context;
import android.content.Intent;

/**
 * This class is core API for libmots library.
 */
public class MessageOnTheScreen {
    private static MessageOnTheScreen sInstance;

    private MessageOnTheScreen(Activity activity, boolean startShowing) {
        Context applicationContext = activity.getApplicationContext();
        applicationContext.startService(new Intent(applicationContext,
                SharedElementService.class));
    }

    /**
     * Returns reference to MessageOnTheScreen instance.
     * @param activity - caller Activity.
     * @param startShowing - if true, then showing of shared element will be
     * started immediately, even when there are no debug messages.
     * @return reference to MessageOnTheScreen instance.
     */
    static synchronized public MessageOnTheScreen getInstance(
            Activity activity, boolean startShowing) {
        if (sInstance == null) {
            sInstance = new MessageOnTheScreen(activity, startShowing);
        }

        show(activity, startShowing);

        return sInstance;
    }

    /**
     * Adds new message to the debug window.
     * @param activity - caller Activity.
     * @param message - text that should be added at the end of the list in the
     * debug window.
     */
    public void addMessage(Activity activity, String message) {
    	Context applicationContext = activity.getApplicationContext();
    	Intent intent = new Intent(applicationContext,
    			SharedElementService.class);
    	intent.putExtra(SharedElementService.COMMAND,
    			SharedElementService.ADD_MESSAGE);
    	intent.putExtra(SharedElementService.NEW_MESSAGE, message);
    	applicationContext.startService(intent);
    }

    /**
     * Sets new text inside debug window (replacing current content).
     * @param activity - caller Activity.
     * @param newText - text that should replace current content of debug
     * window.
     */
    public void setText(Activity activity, String newText) {
    	Context applicationContext = activity.getApplicationContext();
    	Intent intent = new Intent(applicationContext,
    	        SharedElementService.class);
    	intent.putExtra(SharedElementService.COMMAND,
    			SharedElementService.SET_TEXT);
    	intent.putExtra(SharedElementService.NEW_MESSAGE, newText);
    	applicationContext.startService(intent);
    }

    private static void show(Activity activity, boolean startShowing) {
        Context applicationContext = activity.getApplicationContext();
        Intent intent = new Intent(applicationContext,
                SharedElementService.class);
        intent.putExtra(SharedElementService.COMMAND,
            SharedElementService.SHOW_SHARED_ELEMENT);
        intent.putExtra(SharedElementService.START_SHOWING, startShowing);
        applicationContext.startService(intent);
    }
}

src/net/wiktorlawski/messageonthescreen/SharedElementService.java
/*
 * The MIT License (MIT)
 *
 * Copyright (c) 2014 Wiktor Lawski <wiktor.lawski@gmail.com>
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in
 * all copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
 * THE SOFTWARE.
 */

package net.wiktorlawski.messageonthescreen;

import java.util.ArrayList;

import android.app.Service;
import android.content.Context;
import android.content.Intent;
import android.graphics.PixelFormat;
import android.os.Bundle;
import android.os.Handler;
import android.os.IBinder;
import android.os.Vibrator;
import android.view.Gravity;
import android.view.MotionEvent;
import android.view.OrientationEventListener;
import android.view.ViewConfiguration;
import android.view.WindowManager;
import android.view.WindowManager.LayoutParams;
import android.widget.ImageView;

/**
 * Service ensures that interactive element exists at the top of other Activity.
 */
public class SharedElementService extends Service {
    public static final String COMMAND = "COMMAND";
    public static final String NEW_MESSAGE = "NEW_MESSAGE";
    public static final String START_SHOWING = "START_SHOWING";

    public static final int INCORRECT_COMMAND = 0;
    public static final int SET_TEXT = 1;
    public static final int SHOW_SHARED_ELEMENT = 2;
    public static final int ADD_MESSAGE = 3;

    private static final int SHARED_ELEMENT_HEIGHT = 100;
    private static final int SHARED_ELEMENT_WIDTH = 100;
    private static final int SHARED_ELEMENT_PADDING_X = 25;
    private static final int SHARED_ELEMENT_PADDING_Y = 25;

    private ArrayList<String> debugMessages = new ArrayList<String>();
    private OrientationListener orientationListener;
    private SharedElementView sharedElement;
    private LayoutParams sharedElementParameters;

    /* If true, shared element should be always visible */
    private boolean showing;
    private boolean visible; /* Current visibility of shared element */
    private WindowManager windowManager;

    @Override
    public IBinder onBind(Intent intent) {
        return null;
    }

    @Override
    public void onCreate() {
        super.onCreate();

        windowManager = (WindowManager) getSystemService(WINDOW_SERVICE);

        sharedElement = new SharedElementView(this);
        sharedElement.setImageResource(R.drawable.icon);
        sharedElementParameters = new LayoutParams(SHARED_ELEMENT_WIDTH,
                SHARED_ELEMENT_HEIGHT, LayoutParams.TYPE_PHONE,
                LayoutParams.FLAG_NOT_FOCUSABLE, PixelFormat.TRANSPARENT);
        sharedElementParameters.gravity = Gravity.LEFT | Gravity.TOP;
        sharedElementParameters.x =
                getScreenWidth() - SHARED_ELEMENT_WIDTH -
                SHARED_ELEMENT_PADDING_X;
        sharedElementParameters.y = SHARED_ELEMENT_PADDING_Y;

        orientationListener = new OrientationListener();
        orientationListener.enable();
    }

    @Override
    public void onDestroy() {
        super.onDestroy();

        if ((sharedElement != null) && visible) {
            windowManager.removeView(sharedElement);
        }
    }

    /*
     * This is the old onStart method that will be called on the pre-2.0
     * platform.  On 2.0 or later we override onStartCommand() so this
     * method will not be called.
     *
     * Reference:
     * http://developer.android.com/reference/android/app/Service.html
     */
    @Override
    public void onStart(Intent intent, int startId) {
    	handleCommand(intent);
    }
    
    @Override
    public int onStartCommand(Intent intent, int flags, int startId) {
    	handleCommand(intent);

    	/* Continue running until explicitly stopped. */
    	return START_STICKY;
    }

    private void handleCommand(Intent intent) {
    	if (intent == null) {
            return;
    	}

    	Bundle bundle = intent.getExtras();

    	if (bundle == null) {
            return;
    	}

    	int command = bundle.getInt(COMMAND, INCORRECT_COMMAND);
    	String newMessage = bundle.getString(NEW_MESSAGE);
    	boolean startShowing = bundle.getBoolean(START_SHOWING, false);

    	switch (command) {
    	case SET_TEXT:
    	    debugMessages.clear();
    	    /* No break */

    	case ADD_MESSAGE:
    	    if (newMessage != null) {
    	        debugMessages.add(newMessage);
    	        showSharedElement(showing);
    	    } else {
    	    	hideSharedElement(showing);
    	    }

    	    break;

    	case SHOW_SHARED_ELEMENT:
    	    if (startShowing) {
    	    	showSharedElement(startShowing);
    	    } else {
    	    	hideSharedElement(startShowing);
    	    }

    	    break;
    	}
    }

    /* package */ void clearDebugMessages() {
    	debugMessages.clear();

    	if (!showing) {
    		hideSharedElement(showing);
    	}
    }

    @SuppressWarnings("deprecation")
    private int getScreenHeight() {
        return windowManager.getDefaultDisplay().getHeight();
    }

    @SuppressWarnings("deprecation")
    private int getScreenWidth() {
        return windowManager.getDefaultDisplay().getWidth();
    }

    private void hideSharedElement(boolean newShowing) {
    	showing = newShowing;

    	if ((debugMessages.isEmpty() == false) || !visible) {
    		return;
    	}

    	windowManager.removeView(sharedElement);

    	visible = false;
    }

    private void showSharedElement(boolean newShowing) {
    	showing = newShowing;

    	if (visible) {
    		return;
    	}

    	windowManager.addView(sharedElement, sharedElementParameters);

    	visible = true;
    }

    private class OrientationListener extends OrientationEventListener {
        private boolean previouslyPortrait;

        public OrientationListener() {
            super(getApplicationContext());

            previouslyPortrait = isPortrait();
        }

        private void adaptToOrientationChange() {
            float previousWidth = getScreenHeight();
            float previousHeight = getScreenWidth();
            float width = getScreenWidth();
            float height = getScreenHeight();
            float scaleWidth = width / previousWidth;
            float scaleHeight = height / previousHeight;

            sharedElementParameters.x *= scaleWidth;
            sharedElementParameters.y *= scaleHeight;
            windowManager.updateViewLayout(sharedElement,
                    sharedElementParameters);
        }

        /* package */ boolean isPortrait() {
            if (getScreenWidth() < getScreenHeight()) {
                return true;
            }

            return false;
        }

        @Override
        public void onOrientationChanged(int orientation) {
            if (orientation == ORIENTATION_UNKNOWN) {
                return;
            }

            if (isPortrait()) {
                if (!previouslyPortrait) {
                    adaptToOrientationChange();
                    previouslyPortrait = true;
                }
            } else {
                if (previouslyPortrait) {
                    adaptToOrientationChange();
                    previouslyPortrait = false;
                }
            }
        }
    }

    private class SharedElementView extends ImageView {
        private static final int BIN_HEIGHT = 100;
        private static final int BIN_WIDTH = 100;
        private static final float TOUCH_ERROR_MARGIN = 7.5f;

        private ImageView binView;
        private float initialTouchX;
        private float initialTouchY;
        private float lastTouchX = -1;
        private float lastTouchY = -1;
        private boolean longPressed;
        private Handler longPressHandler = new Handler();
        private Runnable longPressRunnable = new LongPressRunnable();
        private boolean moved;
        private boolean readyToRemove;

        public SharedElementView(Context context) {
            super(context);
        }

        private void cancelLongPressBehavior() {
            longPressed = false;
            sharedElement.setImageResource(R.drawable.icon);
            longPressHandler.removeCallbacks(longPressRunnable);

            if (binView != null) {
                windowManager.removeView(binView);
                binView = null;
            }
        }

        private boolean isOverlappingBin() {
            final int showNavigationBarId = getResources()
                    .getIdentifier("config_showNavigationBar", "bool",
                        "android");

            final boolean hasNavigationBar = (showNavigationBarId > 0) ?
                    getResources().getBoolean(showNavigationBarId)
                    : false;

            final int navigationBarHeightPortrait =
                    hasNavigationBar ?
                        getResources().getDimensionPixelSize(
                                getResources()
                                .getIdentifier(
                                        "navigation_bar_height",
                                        "dimen", "android"))
                        : 0;

            final int navigationBarHeight =
                    orientationListener.isPortrait() ?
                        navigationBarHeightPortrait
                        /*
                         * Navigation bar located on the left or right in
                         * landscape mode
                         */
                        : 0;

            final int binX0 = (getScreenWidth() - BIN_WIDTH) / 2;
            final int binX1 = (getScreenWidth() + BIN_WIDTH) / 2;
            final int binY0 =
                    getScreenHeight() - BIN_HEIGHT - navigationBarHeight;

            if ((((sharedElementParameters.x >= binX0) &&
                    (sharedElementParameters.x <= binX1)) ||
                    ((sharedElementParameters.x + SHARED_ELEMENT_WIDTH >=
                            binX0) &&
                    (sharedElementParameters.x + SHARED_ELEMENT_WIDTH <=
                            binX1))) &&
                    (sharedElementParameters.y + SHARED_ELEMENT_HEIGHT >=
                            binY0)) {
                return true;
            }

            return false;
        }

        private boolean isSignificantDifference(float rawX, float rawY) {
            if ((initialTouchX >= rawX - TOUCH_ERROR_MARGIN) &&
                    (initialTouchX <= rawX + TOUCH_ERROR_MARGIN) &&
                    (initialTouchY >= rawY - TOUCH_ERROR_MARGIN) &&
                    (initialTouchY <= rawY + TOUCH_ERROR_MARGIN)) {
                return false;
            }

            return true;
        }

        @Override
        public boolean onTouchEvent(MotionEvent ev) {
            final float x = ev.getRawX();
            final float y = ev.getRawY();

            switch (ev.getAction()) {
            case MotionEvent.ACTION_CANCEL:
                cancelLongPressBehavior();
                readyToRemove = false;

                break;

            case MotionEvent.ACTION_DOWN:
                initialTouchX = lastTouchX = ev.getRawX();
                initialTouchY = lastTouchY = ev.getRawY();
                moved = false;

                cancelLongPressBehavior();
                readyToRemove = false;
                longPressHandler.postDelayed(longPressRunnable,
                        ViewConfiguration.getLongPressTimeout());

                break;

            case MotionEvent.ACTION_MOVE:
                /*
                 * Do not move shared element without previous ACTION_DOWN or
                 * ACTION_UP.
                 */
                if ((lastTouchX != -1) && (lastTouchY != -1)) {
                    sharedElementParameters.x += (int) (x - lastTouchX);
                    sharedElementParameters.y += (int) (y - lastTouchY);
                    windowManager.updateViewLayout(sharedElement,
                        sharedElementParameters);

                    if (isSignificantDifference(x, y)) {
                    	moved = true;

                        if (!longPressed) {
                            cancelLongPressBehavior();
                        }
                    }

                    if (longPressed) {
                        if (this.isOverlappingBin()) {
                            if (!readyToRemove) {
                                sharedElement
                                .setImageResource(R.drawable.icon_stroke_red);
                                readyToRemove = true;
                            }
                        } else if (readyToRemove) {
                            sharedElement
                            .setImageResource(R.drawable.icon_stroke);
                            readyToRemove = false;
                        }
                    }
                }

                lastTouchX = x;
                lastTouchY = y;

                break;

            case MotionEvent.ACTION_UP:
                if (!moved && (isSignificantDifference(x, y) == false) &&
                        !longPressed) {
                    new DebugWindow(getApplicationContext(), debugMessages,
                        SharedElementService.this);
                }

                initialTouchX = lastTouchX = x;
                initialTouchY = lastTouchY = y;
                moved = false;

                cancelLongPressBehavior();

                if (readyToRemove) {
                    SharedElementService.this.stopSelf();
                }

                break;
            }

            return true;
        }

        private class LongPressRunnable implements Runnable {
            private static final long VIBRATE_DURATION_MS = 50;

            @Override
            public void run() {
                longPressed = true;
                sharedElement.setImageResource(R.drawable.icon_stroke);

                Vibrator vibrator =
                        (Vibrator) getSystemService(VIBRATOR_SERVICE);
                vibrator.vibrate(VIBRATE_DURATION_MS);

                createBinView();
            }

            private void createBinView() {
                binView = new ImageView(SharedElementService.this);
                binView.setImageResource(R.drawable.bin);
                LayoutParams binViewParameters = new LayoutParams(BIN_WIDTH,
                        BIN_HEIGHT, LayoutParams.TYPE_PHONE,
                        LayoutParams.FLAG_NOT_FOCUSABLE,
                        PixelFormat.TRANSPARENT);
                binViewParameters.gravity = Gravity.CENTER | Gravity.BOTTOM;

                windowManager.addView(binView, binViewParameters);
            }
        }
    }
}
